lib/DxilPIXPasses/DxilShaderAccessTracking.cpp

```diff
@@ -302,110 +337,401 @@ void DxilShaderAccessTracking::EmitAccess(LLVMContext &Ctx, OP *HlslOP,
+          Constant* B = cast<Constant>(createHandleFromBinding.get_bind());
+          auto binding = hlsl::resource_helper::loadBindingFromConstant(*B);
+          {
+            ret.accessStyle = AccessStyle::FromRootSig;
+            ret.index = createHandleFromBinding.get_index();
+            ret.registerType = RegisterTypeFromResourceClass(
+                static_cast<hlsl::DXIL::ResourceClass>(binding.resourceClass));
+
+            //ret.RegisterID = binding.rangeLowerBound;
+            ret.RegisterSpace = binding.spaceID;
+          }
+      } else if (hlsl::OP::IsDxilOpFuncCallInst(handleCreation, hlsl::OP::OpCode::CreateHandleFromHeap)) {
+          DxilInst_CreateHandleFromHeap createHandleFromHeap(handleCreation);
+          ret.accessStyle = createHandleFromHeap.get_samplerHeap_val()
+              ? AccessStyle::SamplerFromDescriptorHeap : AccessStyle::ResourceFromDescriptorHeap;
+          ret.dynamicallyBoundIndex = createHandleFromHeap.get_index();
+
+          ret.registerType = RegisterTypeFromResourceClass(properties.getResourceClass());
+
+          DynamicResourceBinding drb{};
+          drb.HeapIsSampler = createHandleFromHeap.get_samplerHeap_val();
+          drb.HeapIndex = -1;
+          drb.Name = "ShaderNameTodo";
+          if (auto * constInt = dyn_cast<ConstantInt>(createHandleFromHeap.get_index()))
+          {
+              drb.HeapIndex = constInt->getLimitedValue();
+          }
+          m_dynamicResourceBindings.emplace_back(std::move(drb));
+
+          return ret;
+      } else {
+          assert(false);
+      }
+  }
 
   return ret;
 }
 
+static bool IsDynamicResourceShaderModel(DxilModule& DM) {
+    auto major = DM.GetShaderModel()->GetMajor();
+    auto minor = DM.GetShaderModel()->GetMinor();
+    return major > 6 || (major == 6 && minor >= 6);
+}
+
+// Set up a UAV with structure of a single int
+static llvm::CallInst* CreateUAV(DxilModule & DM, IRBuilder<> & Builder, unsigned int UAVResourceHandle, unsigned int bind, const char * name)
+{
+    LLVMContext& Ctx = DM.GetModule()->getContext();
+
+    SmallVector<llvm::Type*, 1> Elements{ Type::getInt32Ty(Ctx) };
+    llvm::StructType* UAVStructTy =
+        llvm::StructType::create(Elements, "class.RWStructuredBuffer");
+    std::unique_ptr<DxilResource> pUAV =
+        llvm::make_unique<DxilResource>();
+    pUAV->SetGlobalName((std::string("PIX_CountUAVName")+ std::to_string(UAVResourceHandle)).c_str());
+    pUAV->SetGlobalSymbol(UndefValue::get(UAVStructTy->getPointerTo()));
+    pUAV->SetID(UAVResourceHandle);
+    pUAV->SetRW(true); //sets UAV class
+    pUAV->SetSpaceID((
+        unsigned int)-2); // This is the reserved-for-tools register space
+    pUAV->SetSampleCount(1);
+    pUAV->SetGloballyCoherent(false);
+    pUAV->SetHasCounter(false);
+    pUAV->SetCompType(CompType::getI32());
+    pUAV->SetLowerBound(0);
+    pUAV->SetRangeSize(1);
+    pUAV->SetKind(DXIL::ResourceKind::RawBuffer);
```

Comment:
As a future suggestion, this section was particularly hard to identify what had changed and what had just moved. When faced with this, I like to make a single change that moves things without changing them as the first change so that movements can be evaluated separately from changes to moved code

Reply:
Noted. Codeflow does a somewhat better job of showing moved code... too bad we can't use it here.